### PR TITLE
Serializing collections with non-default constructible data

### LINKED ISSUE #2795: Serialization of types without default constructor
Dear HPX developers,

Are there any plans to support serialization of `std::vector<T>` (and similar STL containers), where `T` has no default constructor? This feature exists in `boost::serialization` but not in `hpx::serialization`, as far as I can see.

Thanks in advance,
Igor

### PR DESCRIPTION
**Pull request by AntonBikineev**
This fixes #2795

### CODE REVIEW

File: hpx/runtime/serialization/detail/serialize_collection.hpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                    sizeof(value_type), alignof(value_type)>;
+
+                collection.clear();
+                while (size-- > 0)
+                {
+                    storage_type storage;
+                    value_type& ref = reinterpret_cast<value_type&>(storage);
+                    load_construct_data(ar, &ref, 0);
+                    ar >> ref;
+                    collection.push_back(ref);
```
Comment (hkaiser) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Shouldn't this be

    collection.push_back(std::move(ref));

?

File: hpx/runtime/serialization/detail/serialize_collection.hpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            template <class Archive, class Collection>
+            static void call(Archive& ar, Collection& collection,
+                    typename Collection::size_type size)
+            {
+                using value_type = typename Collection::value_type;
+                using storage_type = std::aligned_storage<
+                    sizeof(value_type), alignof(value_type)>;
+
+                collection.clear();
+                while (size-- > 0)
```
Comment (hkaiser) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Would it be possible to call `collection.reserve(size);` here (at least for `vector` and `deque` etc.)?

File: hpx/runtime/serialization/detail/serialize_collection.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+#include <type_traits>
+
+namespace hpx { namespace serialization { namespace detail {
+
+    // default fallbacks. in fact, they should never be called
+    template <class Archive, class T>
+    void save_construct_data(Archive&, T*, unsigned)
+    {
+    }
```
Comment (hkaiser) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Please add a `HPX_ASSERT(false);` if this function is supposed to never be called.

File: hpx/runtime/serialization/detail/serialize_collection.hpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    template <class Archive, class T>
+    void load_construct_data(Archive&, T* t, unsigned)
+    {
+        ::new (t) T;
```
Comment (hkaiser) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Same here, please add `HPX_ASSERT(false);`

File: hpx/runtime/serialization/detail/serialize_collection.hpp (hunk 2 above)
Comment (AntonBikineev) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Good note, thanks!

File: hpx/runtime/serialization/detail/serialize_collection.hpp (hunk 1 above)
Comment (AntonBikineev) [inline hpx/runtime/serialization/detail/serialize_collection.hpp]: Nice catch, thanks!

**Comment by hkaiser** [comment]
@krivenko could you please verify whether this is solving your issue?

**Comment by hkaiser** [comment]
@AntonBikineev inspect is unhappy: https://7649-4455628-gh.circle-artifacts.com/0/tmp/circle-artifacts.Y4rGbRn/hpx_inspect_report.html

**Comment by hkaiser** [comment]
@AntonBikineev will you have time to look into @krivenko's problems or do you want me to do that?

**Comment by AntonBikineev** [comment]
@hkaiser, yeah, I’m looking into this…

**Comment by hkaiser** [comment]
@krivenko please let us know if you're ok with these changes.

**Comment by hkaiser** [comment]
This was approved by @krivenko here: #2795.

**Comment by AntonBikineev** [comment]
@hkaiser I don't have permissions to merge this PR, could you do this please (if you feel it's ready)?

**Comment by hkaiser** [comment]
@AntonBikineev I will do that. Let's get some of the other, older PRs in first.

**Comment by hkaiser** [comment]
@AntonBikineev Anton, sorry for getting back to this, but why did you use a different mechanism to construct an instance of non-default-constructable types here than we've been using for serializing single instances of such types?

**Comment by AntonBikineev** [comment]
@hkaiser, hm..., it seems like I just looked at what Boost.Serializaiton does. Having the same mechanism for deserializing collections and single instances would be definitely better.

**Comment by hkaiser** [comment]
@AntonBikineev see #2849 for how I propose to unify things. Please have a look.

**Pull request merged**